### PR TITLE
feat: Dbauthz is now default, remove out of experimental

### PR DESCRIPTION
**Pull request by Emyrk**
# What this does

Almost all RBAC checks happen in dbauthz layer now.

# Unit tests

I removed `AuthorizeAllEndpoints`, it was flawed anyway.

A new method of doing rbac assertions was added if you want to check this in a route test. This is no longer required though as `dbauthz` enforces rbac checks. This exists for edge cases or debugging. Tests are optional similar to how we check audit log entries.

### CODE REVIEW

File: coderd/workspaces_test.go [hunk 1]
```diff
@@ -36,8 +36,9 @@ func TestWorkspace(t *testing.T) {
 
 	t.Run("OK", func(t *testing.T) {
 		t.Parallel()
-		client := coderdtest.New(t, &coderdtest.Options{IncludeProvisionerDaemon: true})
+		client, _, api := coderdtest.NewWithAPI(t, &coderdtest.Options{IncludeProvisionerDaemon: true})
 		user := coderdtest.CreateFirstUser(t, client)
+		authz := coderdtest.AssertRBAC(t, api, client)
```
Comment (Emyrk) [inline coderd/workspaces_test.go:41]: See this unit test to see how to asset an rbac action was checked. 

This is a very basic check on 1 `GET`, but shows how you can assert certain rbac checks are run with the given client as the subject.

**Review by johnstcn — CHANGES_REQUESTED** [review]
Blocked by https://github.com/coder/coder/issues/6654

File: codersdk/workspaces.go [hunk 2]
```diff
@@ -22,6 +22,7 @@ type Workspace struct {
 	UpdatedAt                            time.Time      `json:"updated_at" format:"date-time"`
 	OwnerID                              uuid.UUID      `json:"owner_id" format:"uuid"`
 	OwnerName                            string         `json:"owner_name"`
+	OrganizationID                       uuid.UUID      `json:"organization_id" format:"uuid"`
```
Comment (johnstcn) [inline codersdk/workspaces.go:25]: Was this intentional?

File: coderd/workspaces_test.go (hunk 1 above)
Comment (johnstcn) [reply coderd/workspaces_test.go:41]: This is a nice feature to have, but given that it essentially duplicates what we already codify in our authz_querier, we should probably use it sparingly.

File: coderd/workspacebuilds.go [hunk 3]
```diff
@@ -326,7 +311,9 @@ func (api *API) postWorkspaceBuilds(rw http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	// Rbac action depends on the transition
+	// Doing this up front saves a lot of work if the user doesn't have permission.
+	// This is checked again in the dbauthz layer, but the check is cached
+	// and will be a noop later.
```
Comment (johnstcn) [inline coderd/workspacebuilds.go:316]: :+1:

File: coderd/database/dbauthz/querier.go [hunk 4]
```diff
@@ -1068,7 +1068,11 @@ func (q *querier) UpdateUserHashedPassword(ctx context.Context, arg database.Upd
 
 	err = q.authorizeContext(ctx, rbac.ActionUpdate, user.UserDataRBACObject())
 	if err != nil {
-		return err
+		// Admins can update passwords for other users.
+		err = q.authorizeContext(ctx, rbac.ActionUpdate, user.RBACObject())
+		if err != nil {
+			return err
+		}
```
Comment (johnstcn) [inline coderd/database/dbauthz/querier.go:1075]: Good catch; this should probably warrant a separate unit test

File: coderd/coderdtest/coderdtest.go [hunk 5]
```diff
@@ -183,18 +182,18 @@ func NewOptions(t *testing.T, options *Options) (func(http.Handler), context.Can
...
+		options.Authorizer = &RecordingAuthorizer{
+			Wrapped: rbac.NewCachingAuthorizer(prometheus.NewRegistry()),
+		}
+	}
+
 	if options.Database == nil {
 		options.Database, options.Pubsub = dbtestutil.NewDB(t)
-	}
-	// TODO: remove this once we're ready to enable authz querier by default.
-	if strings.Contains(os.Getenv("CODER_EXPERIMENTS_TEST"), string(codersdk.ExperimentAuthzQuerier)) {
```
Comment (johnstcn) [inline coderd/coderdtest/coderdtest.go:190]: 🙌

File: coderd/coderdtest/authorize.go [hunk 6]
```diff
@@ -1,559 +1,144 @@
...
-
-type RouteCheck struct {
-	NoAuthorize  bool
-	AssertAction rbac.Action
-	AssertObject rbac.Object
-	StatusCode   int
+	// We use the database directly to not cause additional auth checks on behalf
+	// of the user. This does add authz checks on behalf of the system user, but
+	// it is hard to avoid that.
+	ctx := dbauthz.AsSystemRestricted(context.Background())
```
Comment (johnstcn) [inline coderd/coderdtest/authorize.go:50]: It works! 

image

File: coderd/coderdtest/authorize.go [hunk 7]
```diff
@@ -1,559 +1,144 @@
...
-	// skipRoutes allows skipping routes from being checked.
-	skipRoutes := map[string]string{
-		"POST:/api/v2/users/logout": "Logging out deletes the API Key for other routes",
-		"GET:/derp":                 "This requires a WebSocket upgrade!",
-		"GET:/derp/latency-check":   "This always returns a 200!",
-	}
+// RBACAsserter is a helper for asserting that the correct RBAC checks are
+// performed. This struct is tied to a given user, and only authorizes calls
+// for this user are checked.
+type RBACAsserter struct {
```
Comment (johnstcn) [inline coderd/coderdtest/authorize.go:29]: This is really nice to have integrated into `coderdtest`.

**Review by johnstcn — COMMENTED** [review]
Looking good so far; just one comment where we should probably add a separate unit test for the case of an admin resetting another user's password.

File: coderd/database/dbauthz/querier.go (hunk 4 above)
Comment (Emyrk) [reply coderd/database/dbauthz/querier.go:1075]: It is here: https://github.com/coder/coder/blob/5b07f1e2a399b9acc568159fdac81b93552deda3/coderd/users_test.go#L617-L617

File: coderd/apikey.go [hunk 8]
```diff
@@ -306,16 +306,24 @@ func (api *API) tokens(rw http.ResponseWriter, r *http.Request) {
 // @Router /users/{user}/keys/{keyid} [delete]
 func (api *API) deleteAPIKey(rw http.ResponseWriter, r *http.Request) {
 	var (
-		ctx   = r.Context()
-		keyID = chi.URLParam(r, "keyid")
+		ctx               = r.Context()
+		keyID             = chi.URLParam(r, "keyid")
+		auditor           = api.Auditor.Load()
+		aReq, commitAudit = audit.InitRequest[database.APIKey](rw, &audit.RequestParams{
```
Comment (johnstcn) [inline coderd/apikey.go:312]: 👀

File: coderd/database/dbauthz/querier.go (hunk 4 above)
Comment (johnstcn) [reply coderd/database/dbauthz/querier.go:1075]: Whoops, I missed it!

File: coderd/apikey.go (hunk 8 above)
Comment (Emyrk) [reply coderd/apikey.go:312]: Huh this came from the merge. Idk why it is a diff :thinking:

Comment (Emyrk) [reply coderd/apikey.go:312]: This is in main :thinking: 

https://github.com/coder/coder/blob/f7842ee1a292e2db767be7f08f145e0b210e99e4/coderd/apikey.go#L308-L319

**Review by johnstcn — APPROVED** [review]

**Pull request merged**